### PR TITLE
✨ Add List-style Dashboard Dropdown Filters (#2001 #2008 #2009)

### PR DESCRIPTION
**Pull request by mistryrn**
**Description of changes**

**TL;DR: creates `ListFilter` inside `DropdownPanel`, a fully accessible checkbox list filter panel, and applies them to the headers of the "Core Completion", "Samples", "Raw Reads", and "Alignment Status" columns in the Dashboard table. Also includes some refactoring and bug fixes for the existing `TextInputFilter` component added previously.**

The biggest addition in this PR is the new `ListFilter` component within `DropdownPanel` of UIKit. It satisfies the design and behaviour requirements of https://github.com/icgc-argo/platform-ui/issues/2001 , which I felt are different enough from the existing `OptionsList` component in UIKit to warrant its own component. This is fully keyboard accessible, and has props for providing specific labels for screen readers if needed.

List filters have been added to the headers of the following columns of the Dashboard table: "Core Completion" (#2001), "Samples" (#2008), "Raw Reads" (#2008), and "Alignment Status" (#2009)

Due to the similarities between the `TextInputFilter` panel and the `ListFilter` panel, a great deal of refactoring was done to try to minimize any code duplication.

Storybook has been updated with examples of each type of dropdown filter, as well as a new generic "Basic" version of the `DropdownPanel` component.

**Type of Change**

- [ ] Bug
- [ ] Styling
- [x] New Feature

**Checklist before requesting review:**

- Design (select one):
  - [x] Matches design:
    - component sizes, spacing, and styles
    - font size, weight, colour
    - spelling has been double checked
  - [ ] No design provided, screenshot of changes included in PR
  - [ ] No changes to UI
- UI Kit (select one):
  - [ ] New Components with storybook stories created
  - [x] Modified Existing components and updates stories
  - [ ] No new UIKit components or changes
- [x] Feature is minimally responsive
- [x] Manual testing of UI feature
- [x] Add copyrights to new files
- [ ] Connected ticket to PR

### CODE REVIEW

File: components/pages/submission-system/program-dashboard/DonorDataSummary/DonorSummaryTable.tsx [hunk 1]
```diff
@@ -232,50 +301,83 @@ export default ({
...
+    onFilter = () => {},
+  }: {
+    header: string;
+    panelLegend?: string;
+    filterOptions?: Array<FilterOption>;
+    activeFilters?: Array<string>;
+    onFilter?: (filters?: Array<FilterOption>) => void;
+  }) => {
+    const [open, setOpen] = useState(false);
+    const options = React.useMemo(
```
Comment (mistryrn) [inline components/pages/submission-system/program-dashboard/DonorDataSummary/DonorSummaryTable.tsx:342]: Generating the options using `useMemo` to avoid re-renders of the child components.

In a future ticket, pending an API update, this will also be used to add the `doc_count` to each option.

File: uikit/DropdownPanel/index.tsx [hunk 2]
```diff
@@ -86,6 +132,58 @@ export const DropdownPanelButtonSection = styled('div')`
   justify-content: space-between;
 `;
 
+export const FilterWrapper = ({
```
Comment (mistryrn) [inline uikit/DropdownPanel/index.tsx:135]: Refactoring (pt 4):
- `FilterWrapper` encapsulates the repeated components of the `DropdownPanel` between the text and list filters: form, fieldset, legend, and cancel/confirm buttons

File: uikit/DropdownPanel/index.tsx [hunk 3]
```diff
@@ -104,13 +202,13 @@ export const TextInputFilter = ({
   confirmLabel?: string;
   inputLabel?: string;
   inputPlaceholder?: string;
-  onCancelClick?: (any?) => void;
-  onConfirmClick?: (any?) => void;
-  onInputChange?: (any?) => void;
+  onCancelClick?: (event?: any) => void;
```
Comment (mistryrn) [inline uikit/DropdownPanel/index.tsx:205]: 😅 missed the prop types here previously 🙈

File: uikit/DropdownPanel/index.tsx [hunk 4]
```diff
@@ -121,56 +219,250 @@ export const TextInputFilter = ({
...
+      onCancelClick={(e) => {
+        e.preventDefault();
+        setOpen(false);
+        onCancelClick();
+      }}
+      onConfirmClick={(e) => {
+        e.preventDefault();
+        onConfirmClick(text?.trim());
+      }}
+      confirmDisabled={!text.length && !initialValue?.length}
```
Comment (mistryrn) [inline uikit/DropdownPanel/index.tsx:235]: Was able to fix an annoying behaviour here. Previously, the "Apply" filter button was disabled whenever the text input was empty. But that prevented you from clearing the filter, so I had it automatically clear inside the onChange whenever the text length went down to zero.

This was really annoying, as whenever you deleted all your input text the panel would immediately close because the filter was applied and the table would re-render.

Now, instead of all that, we just check to see if there is a filter currently applied using the `initialValue` prop. If there is, we are able to use the "Apply" button when there is no text in the input to clear if/when we want to.

File: uikit/DropdownPanel/index.tsx [hunk 5]
```diff
@@ -121,56 +219,250 @@ export const TextInputFilter = ({
...
+          }}
+          ref={_inputRef}
+          showClear={true}
+        />
+      </DropdownPanelInputSection>
+    </FilterWrapper>
+  );
+};
+
+let listFilterInstances = 0;
```
Comment (mistryrn) [inline uikit/DropdownPanel/index.tsx:257]: Real juicy part of the PR is here (the `ListFilter` component).

Quick explainer for the "listFilterInstances" (there's a comment further down where it gets used but just in case):
- it's likely that we'll end up with multiple instances of these per page, so to ensure uniqueness among IDs for the options (but to still keep them predictable), we keep track of the count of instances and prefix the `instanceNum` to each option

File: uikit/DropdownPanel/index.tsx [hunk 6]
```diff
@@ -121,56 +219,250 @@ export const TextInputFilter = ({
...
+    setOptions(options.map((option) => ({ ...option, isChecked: true })));
+    onSelectAllOptions();
+  };
+
+  const clear = () => {
+    setOptions(options.map((option) => ({ ...option, isChecked: false })));
+    onClear();
+  };
+
+  const focusNext = (i) => {
```
Comment (mistryrn) [inline uikit/DropdownPanel/index.tsx:312]: Keyboard a11y stuff specific to the checkbox lists from here to line 349:
- focus on the first checkbox when opening the panel
- can use the up/down arrows to navigate the list in addition to tab
- `home` goes to the first item, `end` goes to the last item
- can "loop" through the list (going "up" from the first takes you to the last, going "down" from the last takes you to the first) for ease of navigation through long lists

**Review by mistryrn — COMMENTED** [review]
Adding some explanations of the refactors and some decisions made with the new components 👇

File: uikit/DropdownPanel/index.tsx (hunk 4 above)
Comment (mistryrn) [reply uikit/DropdownPanel/index.tsx:235]: Bug ticket for the issue this fixes: #2027
